### ISSUE TITLE
Digital Sequencer expander (scene editor)

### ISSUE THREAD

**Issue by S0210** [clone45/voxglitch#92]
Dear Bret,
The Digital Sequencer looks clean and simple. It fits very well into any patch. But I'd like to propose an expander module that assist programming of a specific usage of the module.

![image](https://user-images.githubusercontent.com/34127628/134706376-5ab55f73-9cf4-49d3-aed9-361575dc47b1.png)

The knob on the top allows you to select a step (1-32). The six buttons would allow you to modify the relevant gates (ON/OFF) of the six sequences. The six knobs would similarly allow you to set the target CV for that step/scene. 

The playback is the same as now. (I'll try to create a short video to demonstrate.) It practically allows creation of upto 32 scene. But the editing/programming part would be much simpler with this expander.

Best regards, Andras

**Comment by clone45** [comment]
Hi Andras,

I'm not quite understanding the functionality that you're suggesting.  Have you considered using the 8-Face module to save and load scenes?  (https://library.vcvrack.com/Stoermelder-P1/EightFace)  I find that it does a great job.

**Comment by S0210** [comment]
Hello,

I'm specifically talking about 'scene programming'. Firstly 8FACE is limited to a single module. 8FACE mk2 is more interesting but probably Bidoo Moire is the best in this field. I'm playing with Moire recently and actually I think it's great. But I feel that with a single expander like DSX-1 even your Digital Sequencer could easily excel here.

To compare Moire and Digital Sequencer (&DSX-1) I think there are a few distinct advantages to be highlighted:
1. your great Digital Sequencer can manage 32 scenes (Moire can 16 scenes, 8FACE can 8 or 16 scenes)
2. Digital Sequencer can provide an excellent overview per scene (DSX-1) as well as per parameters using the current bar window (Moire can show by scene only, 8FACE provides no overview at all)
3. after defining scenes in Digital Sequencer it is easy to jump between them with different speed due to the individual STEP input per tracks (no way to do that in Moire or 8FACE)

Obviously there are other fields DSX-1 will not make your Digital Sequence competitive:
- Moire can slide/morph from scene to scene (your DS can't do that, an additional module like Sled is still needed)
- there is some flexibility in the order of scenes in 8FACE (see manual buttons) and also in Moire (complicated) but Pilot has various step orders (currently DS can only step forward)

I know that you have limited time and you need to set your priorities wisely. This may not seem to be the most bombastic proposal but I think it's a 'low hanging fruit'. I hope that my proposal shows at least the great potential your Digital Sequencer beholds.

Regards, Andras

**Comment by S0210** [comment]
[VCV digitalsequencer scene test1.zip](https://github.com/clone45/voxglitch/files/7271348/VCV.digitalsequencer.scene.test1.zip)

Instead of a video to demonstrate I just upload a small patch here. Basically there are three FM VCOs playing their tune with different BPM. The Digital Sequencer is in charge of changing the 'scenes'. That means regular change of the FM VCO feedbacks. During a 'scene' the VCOs are working independently. Some plays 2 notes, 3 notes or even 4 notes. But the overall sound is also controlled from scene to scene. 

DS is capable of handling this task even now. But to properly design a harmonies we need to see (and to experiment with) the values of a single step of multiple tracks at the same time. The proposed DSX-1 serves exactly this goal.

However it is, you have at least another nice little tune for background music. ;)

Cheers, A.

**Comment by clone45** [comment]
OH!  Now I understand!  I also see why the external editor would be helpful.  I'll see what I can do.  I haven't been doing much module programming lately though, so I can't make any promises, but I totally understand your need for this.  Would it also be helpful to have a CV output and gate output on the expansion module?

![image](https://user-images.githubusercontent.com/1373491/135724364-e1fcbcb7-dece-482a-9a72-e1310be6c7d4.png)

(for being able to hear the values you're changing by patching them somewhere)?

**Comment by clone45** [comment]
I'm wondering if this should be a stand-along module. I'm really interested in exploring the ability to switch between "play" mode and "edit" mode.  Play mode would function like a sequencer.  Edit module would output the values on the scene that you're currently editing.  Hmm...

**Comment by clone45** [comment]
Sorry for the stream of consciousness.  One feature that I hope to add that might help in the meantime would be a "freeze" button (and quick-key "f") the stops all sequence playback and has the sequencer always output the values that you're editing.  That, along with the existing quick keys for switching sequences (keys 1-6) might achieve what you're looking for.

**Comment by S0210** [comment]
> OH! Now I understand! I also see why the external editor would be helpful. I'll see what I can do. I haven't been doing much module programming lately though, so I can't make any promises, but I totally understand your need for this. Would it also be helpful to have a CV output and gate output on the expansion module?
> (for being able to hear the values you're changing by patching them somewhere)?

No. I think it's perfectly fine if DSX-1 top knob just 'fixes' the relevant step of DS. So if I want to hear what I'm setting for the scene, I just use the regular DS outputs.
I'm also concerned about the implications of having these outputs on DSX-1. I.e. you need to plug/unplug your cables after setting the scenes from DSX-1 to DS; you need to provide a way to 'save' current settings back to the DS sequences; etc.

> I'm wondering if this should be a stand-along module. I'm really interested in exploring the ability to switch between "play" mode and "edit" mode. Play mode would function like a sequencer. Edit module would output the values on the scene that you're currently editing. Hmm...

I would vote against a stand-alone module. Firstly you'd just recreate Bidoo's Moire. Secondly you had to consider edit/play changes, as well as scene changes. 

> Sorry for the stream of consciousness. One feature that I hope to add that might help in the meantime would be a "freeze" button (and quick-key "f") the stops all sequence playback and has the sequencer always output the values that you're editing. That, along with the existing quick keys for switching sequences (keys 1-6) might achieve what you're looking for.

I think 'f' would be a wonderful addition. I really like the idea. It is not only useful for scene editing, but for finding CV sweet spots in general.

Best regards, A.

**Comment by S0210** [comment]
> I'm really interested in exploring the ability to switch between "play" mode and "edit" mode. 

I elegantly skipped commenting this sentence above. But this is something I want to reflect on. It is definitely **one of the most exciting points**! (Not only for Digital Sequencer but in general for most VCV modules!)

Cheers, A.

**Comment by jpnielsen** [comment]
maybe an expander could be used for MIDI mapping - if it has a MIDI-mappable knob for each fader.

**Comment by clone45** [comment]
@jpnielsen I really like this idea.  Can you walk me through the user-case for needing MIDI control over the faders?  I can't quite figure out why you wouldn't just convert the MIDI to CV (using, for example, the VCV MIDI CC to CV module), then route that into your modules.  Thanks!!

**Comment by clone45** [comment]
Oh, sorry, was this for Digital Sequencer, or Digital Programmer?

**Comment by jpnielsen** [comment]
I was thinking about the Digital Programmer, so I guess i added to the wrong issue.

With the 16n, I'd want to make a soft-takeover patch  or ask someone to make a module with midi-in and soft takeover.
https://community.vcvrack.com/t/midi-cc-to-cv-selection-s/15546

Digital Programmer seems perfect for the on-screen representation of the faders.
